store/postgres/src/chain_store.rs

```diff
@@ -1159,67 +1168,83 @@ impl ChainStoreTrait for ChainStore {
+    async fn attempt_chain_head_update(
+        self: Arc<Self>,
+        ancestor_count: BlockNumber,
+    ) -> Result<Option<H256>, Error> {
         use public::ethereum_networks as n;
 
         let (missing, ptr) = {
-            let conn = self.get_conn()?;
-            conn.transaction(|| -> Result<(Vec<H256>, Option<(String, i64)>), Error> {
-                let candidate = self.storage.chain_head_candidate(&conn, &self.chain)?;
-                let (ptr, first_block) = match &candidate {
-                    None => return Ok((vec![], None)),
-                    Some(ptr) => (ptr, 0.max(ptr.number.saturating_sub(ancestor_count))),
-                };
-
-                let missing = self.storage.missing_parents(
-                    &conn,
-                    &self.chain,
-                    first_block as i64,
-                    ptr.hash_as_h256(),
-                    self.genesis_block_ptr.hash_as_h256(),
-                )?;
-                if !missing.is_empty() {
-                    return Ok((missing, None));
-                }
+            let chain_store = self.clone();
+            self.pool
+                .with_conn(move |conn, _| {
+                    let candidate = chain_store
+                        .storage
+                        .chain_head_candidate(&conn, &chain_store.chain)
+                        .map_err(CancelableError::from)?;
+                    let (ptr, first_block) = match &candidate {
+                        None => return Ok((None, None)),
+                        Some(ptr) => (ptr, 0.max(ptr.number.saturating_sub(ancestor_count))),
+                    };
+
+                    match chain_store
+                        .storage
+                        .missing_parents(
+                            &conn,
+                            &chain_store.chain,
+                            first_block as i64,
+                            ptr.hash_as_h256(),
+                            chain_store.genesis_block_ptr.hash_as_h256(),
+                        )
+                        .map_err(CancelableError::from)?
+                    {
+                        Some(missing) => {
+                            return Ok((Some(missing), None));
+                        }
+                        None => { /* we have a complete chain, no missing parents */ }
+                    }
 
-                let hash = ptr.hash_hex();
-                let number = ptr.number as i64;
-                update(n::table.filter(n::name.eq(&self.chain)))
-                    .set((
-                        n::head_block_hash.eq(&hash),
-                        n::head_block_number.eq(number),
-                    ))
-                    .execute(&conn)?;
-                Ok((missing, Some((hash, number))))
-            })
-        }?;
+                    let hash = ptr.hash_hex();
+                    let number = ptr.number as i64;
+
+                    conn.transaction(
```

Comment:
Why is an explicit transaction necessary here if a single query is executed?

Reply:
It's not - it's more to make sure that if we ever add more write operations here we have it in a txn.